test/add/sdk/sdk_test.dart

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS d.file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:path/path.dart' as p;
+
+import 'package:test/test.dart';
+
+import '../../descriptor.dart' as d;
+import '../../test_pub.dart';
+
+void main() {
+  setUp(() async {
+    await servePackages((builder) {
+      builder.serve('bar', '1.0.0');
+    });
+
+    await d.dir('flutter', [
+      d.dir('packages', [
+        d.dir('foo', [
+          d.libDir('foo', 'foo 0.0.1'),
+          d.libPubspec('foo', '0.0.1', deps: {'bar': 'any'})
+        ])
+      ]),
+      d.dir('bin/cache/pkg', [
+        d.dir(
+            'baz', [d.libDir('baz', 'foo 0.0.1'), d.libPubspec('baz', '0.0.1')])
+      ]),
+      d.file('version', '1.2.3')
+    ]).create();
+  });
+
+  test("gets an SDK dependency's dependencies", () async {
+    await d.appDir({}).create();
+    await pubAdd(
+        args: ['foo', '--sdk', 'flutter'],
+        environment: {'FLUTTER_ROOT': p.join(d.sandbox, 'flutter')});
+
+    await d.dir(appPath, [
+      d.pubspec({
+        'name': 'myapp',
+        'dependencies': {
+          'foo': {'sdk': 'flutter', 'version': '^0.0.1'}
+        }
+      }),
+      d.packagesFile({
+        'myapp': '.',
+        'foo': p.join(d.sandbox, 'flutter', 'packages', 'foo'),
+        'bar': '1.0.0'
+      })
+    ]).validate();
+  });
+
+  test(
+      "gets an SDK dependency's dependencies with version constraint specified",
+      () async {
+    await d.appDir({}).create();
+    await pubAdd(
+        args: ['foo:0.0.1', '--sdk', 'flutter'],
```

Comment:
I think we need to forbid version constraint when specifying `--sdk <sdk>`.

When a package is installed from the SDK, it is versioned through the `environment.<sdk>` constraint in `pubspec.yaml`.
Ie. it's versioned through the SDK constraint your package/app has.

So specifying a version constraint on a package that whose source is `sdk` never makes sense.
I think we need to special case this to exit with an error message in `pub add`.

Reply:
My bad this is for some reason a supported thing:
https://dart.dev/tools/pub/dependencies#sdk